include/iarc7_fc_comms/CommonFcComms.hpp

```diff
@@ -94,90 +113,165 @@ CommonFcComms<T>& CommonFcComms<T>::getInstance()
+    if (status != FcCommsReturns::kReturnOk) {
+        ROS_ERROR("Failed to retrieve flight controller status");
+        return status;
+    }
     fc.armed = temp_armed;
     fc.auto_pilot = temp_auto_pilot;
     fc.failsafe = temp_failsafe;
-    
-    std_msgs::Float32 battery;
-    #pragma GCC warning "TODO handle failure"
-    flightControlImpl_.getBattery(battery.data);
-
     ROS_DEBUG("Autopilot_enabled: %d", fc.auto_pilot);
     ROS_DEBUG("Armed: %d", fc.armed);
     status_publisher.publish(fc);
+
+    std_msgs::Float32 battery;
+    status = flightControlImpl_.getBattery(battery.data);
+    if (status != FcCommsReturns::kReturnOk) {
+        ROS_ERROR("Failed to retrieve flight controller battery info");
+        return status;
+    }
     ROS_DEBUG("Battery level: %f", battery.data);
     battery_publisher.publish(battery);
 
-
     double attitude[3];
-    flightControlImpl_.getAttitude(attitude);
+    status = flightControlImpl_.getAttitude(attitude);
+    if (status != FcCommsReturns::kReturnOk) {
+        ROS_ERROR("Failed to retrieve attitude from flight controller");
+        return status;
+    }
     ROS_DEBUG("Attitude: %f %f %f", attitude[0], attitude[1], attitude[2]);
     sendOrientationTransform(attitude);
+
+    return FcCommsReturns::kReturnOk;
+}
+
+// Attempt to reconnect, blocking until successful
+template<class T>
+void CommonFcComms<T>::reconnect(const ros::Duration& timeout) {
+    FcCommsReturns status;
+    ros::Time start_time = ros::Time::now();
+    do {
+        status = flightControlImpl_.disconnect();
+    } while (status != FcCommsReturns::kReturnOk
+          && (timeout == ros::Duration(0) || ros::Time::now() - start_time < timeout));
+
+    do {
+        status = flightControlImpl_.connect();
+    } while (status != FcCommsReturns::kReturnOk
+          && (timeout == ros::Duration(0) || ros::Time::now() - start_time < timeout));
 }
 
 // Update the sensors on the flight controller
 template<class T>
 void CommonFcComms<T>::updateSensors(const ros::TimerEvent&)
 {
-
     ros::Time times = ros::Time::now();
 
+    FcCommsReturns status;
 
     // Do different things based on the current connection status.
     switch(flightControlImpl_.getConnectionStatus())
     {
         case FcCommsStatus::kDisconnected:
             ROS_WARN("FC_Comms disconnected");
-            // We don't care about the return value we just reconnect.
-            (void)flightControlImpl_.connect();
+            reconnect();
             break;
-        
+
         case FcCommsStatus::kConnected:
-            #pragma GCC warning "TODO handle failure"
-            flightControlImpl_.handleComms();
-            publishTopics();
+            status = flightControlImpl_.handleComms();
+            while (status != FcCommsReturns::kReturnOk) {
+                reconnect();
+                status = flightControlImpl_.handleComms();
+            }
+
+            while (have_new_direction_command_message_) {
```

Comment:
We can't prove anything, we aren't running on an RTOS so nothing is guaranteed to run ever.

There's no upper limit on how many orientation requests we want to send, so that's only limited by how fast the flight controller can send them without affecting performance, which we'll need to test.  Similar problem for the commands, although that's also dictated by how fast the rest of the system runs.

The only reason I wanted this tested before merging is because I don't want to break the master branch.  That probably means testing the current version at 5Hz before merging, then doing the rest of the testing later and making another PR.